src/frontend/src/instance.rs

```diff
@@ -564,15 +580,31 @@ impl SqlQueryHandler for Instance {
         query: &str,
         query_ctx: QueryContextRef,
     ) -> Vec<server_error::Result<Output>> {
-        match parse_stmt(query)
+        let query_interceptor = self.plugins.get::<SqlQueryInterceptorRef>();
+        let query = match query_interceptor.pre_parsing(query, query_ctx.clone()) {
+            Ok(q) => q,
+            Err(e) => return vec![Err(e)],
+        };
+
+        match parse_stmt(query.as_ref())
             .map_err(BoxedError::new)
             .context(server_error::ExecuteQuerySnafu { query })
         {
             Ok(stmts) => {
                 let mut results = Vec::with_capacity(stmts.len());
                 for stmt in stmts {
-                    match self.query_statement(stmt, query_ctx.clone()).await {
-                        Ok(output) => results.push(Ok(output)),
+                    match query_interceptor.post_parsing(stmt, query_ctx.clone()) {
```

Comment:
`post_parsing` is called multiple times. Maybe it's better to parse all stmts, call `post_parsing`(with stmts), and then execute stmts one by one.

Reply:
Reasonable, since these statements are parsed together, batching them can provide a context. :+1: